warehouse/integrations/vulnerabilities/__init__.py

```diff
@@ -0,0 +1,83 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import List
+
+from warehouse import integrations
+
+
+class InvalidVulnerabilityReportRequest(Exception):
+    def __init__(self, message, reason):
+        self.reason = reason
+        super().__init__(message)
+
+
+class VulnerabilityReportRequest:
+    def __init__(
+        self,
+        project: str,
+        versions: List[str],
+        vulnerability_id: str,
+        advisory_link: str,
+        aliases: List[str],
+    ):
+        self.project = project
+        self.versions = versions
+        self.vulnerability_id = vulnerability_id
+        self.advisory_link = advisory_link
+        self.aliases = aliases
+
+    @classmethod
+    def from_api_request(cls, request):
+        if not isinstance(request, dict):
+            raise InvalidVulnerabilityReportRequest(
+                f"Record is not a dict but: {str(request)[:100]}", reason="format"
+            )
+
+        missing_keys = sorted(
+            {"project", "versions", "id", "link", "aliases"} - set(request)
+        )
+        if missing_keys:
+            raise InvalidVulnerabilityReportRequest(
+                f"Record is missing attribute(s): {', '.join(missing_keys)}",
+                reason="format",
+            )
+
+        return cls(
+            request["project"],
+            request["versions"],
+            request["id"],
+            request["link"],
+            request["aliases"],
+        )
```

Comment:
I'd suggest either
```suggestion
        return cls(
            project=request["project"],
            versions=request["versions"],
            id=request["id"],
            link=request["link"],
            aliases=request["aliases"],
        )
```
Or
```suggestion
        return cls(**request)
```
Overall, relying on the order of parameters when there are more than one parameter usually leads to strange and hard-to-debug bugs :)

Reply:
done